jl4-lsp/src/LSP/L4/Rules.hs

```diff
@@ -739,3 +727,56 @@ rangeOfResolveWarning = \ case
     srcSpanToLspRange Nothing
   Resolve.Ambiguous name _ ->
     srcRangeToLspRange $ rangeOf name
+
+-- ----------------------------------------------------------------------------
+-- Helpers for implementing syntax highlighting
+-- ----------------------------------------------------------------------------
+
+applyPositionMapping :: [SemanticToken] -> PositionMapping -> [SemanticToken]
+applyPositionMapping semTokens positionMapping =
+  Maybe.mapMaybe
+    ( \t ->
+        case toCurrentPosition positionMapping t.start of
+          Nothing -> Nothing
+          Just newPos -> Just (t & #start .~ newPos)
+    )
+    semTokens
+
+-- | @'semanticTokensUsing' phases@
+--
+-- Helper function for defining multi-phase semantic syntax highlighting.
+--
+-- Each phase can produce '[SemanticToken]'s and 'PositionMapping' if the result is outdated.
+-- Tokens obtained from earlier phases take precedence over tokens from later phases.
+--
+-- TODO: this function is inefficient, as all phases are always executed.
+-- It would be better if we could tell whether the result is actually stale and then
```

Comment:
is there a reason why we cannot? can't you check whether the rule currently succeeds and then use the stale result if it doesn't?

Reply:
I think it is possible, but requires a new primitive. A combination of `use` and `useWithStale` also works, but is likely a bit more expensive.

Reply:
I implemented it in terms of `use` and `useWithStale`.